File: hosts/elixir/test/conformance_test.exs
defmodule ConformanceTest do
  # Runs the shared conformance fixtures through the ModelLanguage host and checks
  # them against the same expectations the TypeScript runner uses — proving the
  # Elixir bridge is behaviourally identical to the canonical engine.
  use ExUnit.Case, async: false

  @cases_dir Path.expand("../../../conformance/cases", __DIR__)

  defp codes(list) when is_list(list), do: Enum.map(list, & &1["code"])

  test "conformance parity with TypeScript" do
    for path <- Path.wildcard(Path.join(@cases_dir, "*.json")) do
      f = File.read!(path) |> Jason.decode!()
      expect = f["expect"]
      template = Map.get(f, "template", "")

      if Map.has_key?(expect, "output") do
        options =
          %{}
          |> then(&if Map.has_key?(f, "now"), do: Map.put(&1, "now", f["now"]), else: &1)
          |> then(&if Map.has_key?(f, "snippets"), do: Map.put(&1, "snippets", f["snippets"]), else: &1)

        res =
          ModelLanguage.render(template,
            data: Map.get(f, "data", %{}),
            schema: Map.get(f, "schema", []),
            options: options
          )

        assert res["text"] == expect["output"], "output: #{f["name"]}"
        assert codes(res["warnings"]) == Map.get(expect, "warnings", []), "warnings: #{f["name"]}"
      end

      if Map.has_key?(expect, "diagnostics") do
        res = ModelLanguage.validate(template, schema: Map.get(f, "schema", []))
        got = res["diagnostics"] |> codes() |> Enum.sort()
        assert got == Enum.sort(expect["diagnostics"]), "diagnostics: #{f["name"]}"
      end
    end
  end
end

File: hosts/ruby/test/conformance_test.rb
# frozen_string_literal: true

# Runs the shared conformance fixtures through the Ruby host and checks them
# against the same expectations the TypeScript runner uses — proving the Ruby
# bridge is behaviourally identical to the canonical engine.

require "minitest/autorun"
require "json"
require_relative "../lib/model_language"

class ConformanceTest < Minitest::Test
  CASES_DIR = File.expand_path("../../../conformance/cases", __dir__)
  ENGINE = ModelLanguage::Engine.new

  Dir[File.join(CASES_DIR, "*.json")].sort.each do |path|
    fixture = JSON.parse(File.read(path))
    method_name = "test_#{fixture["name"].gsub(/\W+/, "_")}"

    define_method(method_name) do
      expect = fixture["expect"]

      if expect.key?("output")
        options = {}
        options["now"] = fixture["now"] if fixture.key?("now")
        options["snippets"] = fixture["snippets"] if fixture.key?("snippets")
        res = ENGINE.render(fixture["template"], data: fixture.fetch("data", {}),
                                                 schema: fixture["schema"], options: options)
        assert_equal expect["output"], res["text"]
        assert_equal expect.fetch("warnings", []), res["warnings"].map { |w| w["code"] }
      end

      if expect.key?("diagnostics")
        res = ENGINE.validate(fixture["template"], schema: fixture["schema"])
        assert_equal expect["diagnostics"].sort, res["diagnostics"].map { |d| d["code"] }.sort
      end
    end
  end
end
